### DISCUSSION TITLE
local bitcoind provider, getting error "No wallet is loaded."

### DISCUSSION THREAD

**Discussion by krupan** [1200wd/bitcoinlib#391] [Q&A] [answered]
1. Using BitcoinLib 0.6.14

2. Added a provider for my bitcoind instance in providers.json:

```
  "bitcoind": {
    "provider": "bitcoind",
    "network": "bitcoin",
    "client_class": "BitcoindClient",
    "provider_coin_id": "",
    "url": "http://bitcoinrpc:<password>@localhost:8332",
    "api_key": "",
    "priority": 11,
    "denominator": 100000000,
    "network_overrides": null
  },
```

4. Added rpcuser and rpcpassword to my bitcoin.conf file.

5. Added `debug=rpc` to my bitcoin.conf file.

6. Restarted bitcoind

Using clw and running this command:
```
clw <wallet name> --update-transactions
```

I can see the rpc queries in the bitcoind log debug.log file.  However, when it gets to this query:

```
  2024-03-22T06:04:45Z [rpc] ThreadRPCServer method=getaddressinfo user=bitcoinrpc
```

It fails in bitcoinlib with this message (which I can only see if I add a breakpoint() to services.py on line 193, like so:
```
                _logger.debug("Executed method %s from provider %s" % (method, sp))
                self.resultcount += 1
            except Exception as e:
###  added breakpoint here ###
                breakpoint()

                if not isinstance(e, AttributeError):
                    try:
                        err = e.msg
                    except AttributeError:
                        err = e
                    self.errors.update(
                        {sp: err}
                    )
```

and step through the code, otherwise the error message gets swallowed.

```
  bitcoinlib.services.authproxy.JSONRPCException: -18: No wallet is loaded. Load a wallet using loadwallet or create a new one with createwallet. (Note: A default wallet is no longer automatically created)
```

**Comment by mccwdev** [comment]
The message indicates you have not created a wallet in bitcoind.

Bitcoind and bitcoinlib have separate wallets, they are not automatically connected.

Bitcoinlib normally uses a bitcoind node only to send transactions or get block and transaction information.

**Reply by krupan** [reply]
I understand that Bitcoinlib normally uses a bitcoind node only to send transactions or get block and transaction information.  Isn't that exactly what I'm doing here?  When I run this command:
```
clw <wallet name> --update-transactions
```
I'm asking clw to use bitcoind to get transaction information.  The wallet that I'm naming is a bitcoinlib wallet.  It feels like a clw or bitcoinlib bug if it's looking for a bitcoind wallet here, doesn't it?

**Reply by mccwdev** [reply, marked as answer]
Problem is that bitcoind only has utxo information for it own keys, so if an address is not in it's wallet it does not have utxo / transaction information for that address.

You could counter this by using bitcoind's importprivkey, and import all keys of your bitcoinlib wallet to the bitcoind wallet. Only bitcoind needs to reindex the whole blockchain, after an address import, which can takes up to 30 minutes.

You could also use a bcoin node, which allow for address searches on the blockchain.

**Reply by krupan** [reply]
I see now.  I wasn't aware of that about bitcoind, thanks for explaining!  I'll check out bcoin
